src/main/java/com/emc/pravega/perf/PerfStats.java

```diff
@@ -118,44 +100,62 @@ public synchronized void printAll() {
         */
     }
 
-    public synchronized void printTotal() {
-        long elapsed = System.currentTimeMillis() - start;
-        double recsPerSec = 1000.0 * count / (double) elapsed;
+    public synchronized void printTotal(long endTime) {
+        long elapsed = endTime - start;
+        double recsPerSec = 1000.0 * iteration / (double) elapsed;
         double mbPerSec = 1000.0 * this.bytes / (double) elapsed / (1024.0 * 1024.0);
-        long[] percs = percentiles(this.latencies, 0.5, 0.95, 0.99, 0.999);
-        System.out.printf(
-                "%d records sent, %f records/sec (%.5f MB/sec), %.2f ms avg latency, %.2f ms max " + "latency, %.2f " +
-                        "ms 50th, %.2f ms 95th, %.2f ms 99th, %.2f ms 99.9th.\n",
-                count, recsPerSec, mbPerSec, totalLatency / ((double) count * 1000.0), (double) maxLatency / 1000.0,
-                percs[0] / 1000.0, percs[1] / 1000.0, percs[2] / 1000.0, percs[3] / 1000.0);
+        //double[] percs = percentiles(this.latencies, 0.5, 0.95, 0.99, 0.999);
         System.out.printf(
-                " %s FINAL:, %d, %.5f MB/sec, %.2f, %.2f, %.2f, %.2f, %.2f, %.2f\n", this.action,
-                messageSize, mbPerSec, totalLatency / ((double) count * 1000.0), (double) maxLatency / 1000.0,
-                percs[0] / 1000.0, percs[1] / 1000.0, percs[2] / 1000.0, percs[3] / 1000.0);
+                "%d records %s, %.3f records/sec, %d bytes record size, %.3f MB/sec, %.4f ms avg latency, %.4f ms max latency\n", 
+                iteration, action, recsPerSec,  messageSize, mbPerSec, totalLatency / ((double) count), (double) maxLatency);
+        /*  
+        System.out.printf("latencies percentiles:  %.4f ms 50th, %.4f ms 95th, %.4f ms 99th, %.4f ms 99.9th.\n",
+                           percs[0], percs[1], percs[2], percs[3]);
+        */
+        
     }
 
-    private long[] percentiles(long[] latencies, double... percentiles) {
-        long size = Math.min(count, latencies.length);
-        Arrays.sort(latencies, 0, (int) size);
-        long[] values = new long[percentiles.length];
+    private double[] percentiles(double[] latencies, double... percentiles) {
+        Arrays.sort(latencies, 0, (int)count);
+        double[] values = new double[percentiles.length];
         for (int i = 0; i < percentiles.length; i++) {
-            int index = (int) (percentiles[i] * size);
+            int index = (int) (percentiles[i] * count);
             values[i] = latencies[index];
         }
         return values;
     }
 
-    public CompletableFuture runAndRecordTime(Supplier<CompletableFuture> fn, long startTime, int length, Executor executor) {
-        int iter = this.iteration++;
+    public CompletableFuture runAndRecordTime(Supplier<CompletableFuture> fn, long startTime, int length) throws Exception {
```

Comment:
This method takes a Supplier and then immediately call get on it and assigns it to retVal.
This is needlessly complicated. It could just take retVal. There would be no difference.

Reply:
here , the sendFunction is override in WriterWorker and TransactionWriterWorker ; hence the supplier is required.  the writeAndRecordTime is invoked by both WriterWorker and TransactionWriterWorker with different sendFunction implementation. 
so, supplier required here.   even though it may look complicated, but supplier is anyway required to pass the lambda function to another method.. please let me know if i am missing anything here

Reply:
hi Tom, 
    yes I got your point, but this function runAndRecordTime similar to writeAndRecordTime
which actually takes the Supplier , invocation is the writeevent , its similar to that; i am trying to maintain the same structure here.